src/preprocessing/passes/skeleton_preprocessing.cpp

```diff
@@ -0,0 +1,483 @@
+ #include <cryptominisat5/cryptominisat.h>
+ #endif
+
+ #include "prop/registrar.h"
+ #include "prop/cnf_stream.h"
+ #include "prop/sat_solver_factory.h"
+
+ #include "proof/proof_manager.h"
+
+ #include "context/cdo.h"
+ #include "options/proof_options.h"
+ #include "smt/smt_statistics_registry.h"
+ #include "theory/theory_model.h"
+
+#include <vector>
+
+using namespace CVC4;
+using namespace CVC4::theory;
+
+namespace CVC4 {
+namespace preprocessing {
+namespace passes {
+
+
+/* -------------------------------------------------------------------------- */
+
+SkeletonPreprocessing::SkeletonPreprocessing(PreprocessingPassContext* preprocContext)
+    : PreprocessingPass(preprocContext, "non-clausal-simp-sat")
+{
+}
+
+PreprocessingPassResult SkeletonPreprocessing::applyInternal(
+    AssertionPipeline* assertionsToPreprocess)
+{
+
+  Assert(!options::unsatCores() && !options::fewerPreprocessingHoles());
+
+  d_preprocContext->spendResource(options::preprocessStep());
+
+
+  // Convert the original formula into a boolean formula and solve it
+  // with cryptominisat
+  SatSolver* d_satSolver = SatSolverFactory::createCryptoMinisat(smtStatisticsRegistry(), "non_clausal_simp_solver");
+  CVC4::prop::NullRegistrar d_registrar;
+  context::Context d_context;
+  CVC4::prop::TseitinCnfStream d_cnfStream (d_satSolver, &d_registrar,
+                                            &d_context, true,
+                                            "toCNF-skeleton-preprocessing");
+
+  Trace("skeleton-preprocessing") << "asserting to sat solver" << std::endl;
+
+  unsigned substs_index = d_preprocContext->getSubstitutionsIndex();
+  for (size_t i = 0, size = assertionsToPreprocess->size(); i < size; ++i)
+  {
+    Assert(Rewriter::rewrite((*assertionsToPreprocess)[i])
+              == (*assertionsToPreprocess)[i]);
+       // Don't reprocess substitutions
+       if (substs_index > 0 && i == substs_index)
+       {
+         continue;
+       }
+       Trace("skeleton-preprocessing")
+           << "asserting " << (*assertionsToPreprocess)[i] << std::endl;
+       d_cnfStream.convertAndAssert((*assertionsToPreprocess)[i], false, false,
+                                    RULE_GIVEN);
+  }
+
+
+
+  SatValue result = d_satSolver->solve();
+
+  if (result==SAT_VALUE_FALSE){
+    // If in conflict, just return false
+    Trace("skeleton-preprocessing")
+        << "conflict in non-clausal propagation" << std::endl;
+    Assert(!options::unsatCores() && !options::fewerPreprocessingHoles());
+    assertionsToPreprocess->clear();
+    Node n = NodeManager::currentNM()->mkConst<bool>(false);
+    assertionsToPreprocess->push_back(n);
+    PROOF(ProofManager::currentPM()->addDependence(n, Node::null()));
+    delete d_satSolver;
+    return PreprocessingPassResult::CONFLICT;
+  }
+
+  std::vector<SatLiteral> topLevelUnits = d_satSolver->getTopLevelUnits();
+
+  std::vector<TNode> learned_literals;
+
+
+  for (size_t i = 0; i < topLevelUnits.size(); i++){
```

Comment:
Cache size `for (size_t i = 0, size = topLevelUnits.size(); i < size; ++i) ...`.